### ISSUE TITLE
ocaml-git is using too much memory

### ISSUE THREAD

**Issue by Chris00** [mirage/ocaml-git#125]
This correspond to the [following Irmin bug](https://github.com/mirage/irmin/issues/263).  Code (`c.ml`):

``` ocaml
open Lwt

module Store = Git_unix.FS
module G = Git_unix.Sync.Make(Store)

let main uri =
  Store.create ~root:"/tmp/c.dir" () >>= fun t ->
  G.fetch t uri ~update:true >>= fun res ->
  match Git_unix.Sync.Result.head_contents res with
   | Some h -> Store.write_head t h
   | None -> return_unit

let () =
  let uri = Git.Gri.of_string "git://github.com/xapi-project/xen-api.git" in
  Lwt_main.run(main uri)

;;
(* Local Variables: *)
(* compile-command: "ocamlfind ocamlopt -o c.native -package git.unix -linkpkg c.ml" *)
(* End: *)
```

**Comment by Chris00** [comment]
ocaml-git 1.7.1

**Comment by yallop** [comment]
I spent a while looking at this.  First, it's easily reproducible; `/usr/bin/time -v` reports

```
	Maximum resident set size (kbytes): 2401108
```

i.e. 2.3GB.

Digging about (using @kayceesrk's [allocation profiler](http://kcsrk.info/ocaml/profiling/2015/09/23/bytecode-allocation-profiler/)) reveals that most of the memory is allocated in the `Buffer` and `Bytes` modules.  Further investigation reveals that the [raw](https://github.com/mirage/ocaml-git/blob/e0ef97b8320cdaea54852c9533fe656a5b383a54/lib/packed_value.ml#L139) cache field of the `Packed_value.PIC.t` type is responsible for the majority of resident memory.  Removing the `raw` field altogether, and instead recomputing its value on each call to [unpack](https://github.com/mirage/ocaml-git/blob/e0ef97b8320cdaea54852c9533fe656a5b383a54/lib/packed_value.ml#L173-L179) brings the memory usage down significantly:

```
	Maximum resident set size (kbytes): 954948
```

i.e. 0.91GB.

Unfortunately, it also slows the program down significantly (at least on a machine with plenty of memory), from 5.42s user time to 9.06s.

I experimented a bit with caching the result of calling `unpack`, using the [LRU cache](https://github.com/mirage/ocaml-git/blob/e0ef97b8320cdaea54852c9533fe656a5b383a54/lib/LRU.ml) that's already in the repo.  70% of calls use an argument that should be cachable with a cache size of around 1000 (i.e. that last appeared within the last 1000 or so calls) and 90% of calls with a cache size of around 5000.  Caching brings the time down a bit: with a cache size of 1000 the time decreases from 9.06s to 8.24s, without much effect on memory usage, and with a cache size of 5000 the time decreases from to 6.95s, but with an increase in memory to 1.32GB.

Summary: there's a memory/time tradeoff, at least without more extensive changes.  Options: 2.3GB+5.42s (caching everything), 1.32GB+6.95s (LRU cache size 5000), 0.91GB+8.24s (LRU cache size 1000).

**Comment by yallop** [comment]
> Removing the raw field altogether, and instead recomputing its value on each call to unpack brings the memory usage down significantly:
> 
>     Maximum resident set size (kbytes): 954948
> 
> i.e. 0.91GB.

It's sometimes down as low as 0.68GB.

**Comment by samoht** [comment]
Should be fixed by ocaml-git 2.0
